### PR TITLE
ddl/index: refactor add index backfill data flow

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
Add index use to backfill index by batch worker in each turn.  Specifically, send batch tasks to workers, and then wait this batch tasks finish. Then start the next turn. 

This PR change the backfill flow, as you can see, after send a batch tasks, we don't need to wait all the tasks finish, if some worker finish a task early, the worker can continue to do the next task.

There is one thing should be careful,  we need to store `reorgInfo` , and the `reorgInfo.NextHandle` should be sequential.  eg: if we have 3 tasks: task1, task2, task3, if task2 is finished before task1, we can't store the `reorgInfo.NextHandle = task2.nextHandle`, otherwise, if task1 panic, then ddl worker recover to backfill index, it will start from `task2.nextHandle`, and will jump over the task1. This is not right and will cause inconsistency of index and record.

### What is changed and how it works?
The main code frame is like below, This is from @tiancaiamao , Thanks.
```go
type task struct {
	...
	sync.WaitGroup
}
func addPhysicalTableIndex() {
	go func() {
		defer close(doingTaskCh)
		for _, task := range tasks {
			task.Add(1);
			worker := <- availableWorkerCh
			worker.taskCh <- task
			doingTaskCh <- task 
		}
	}()
	
	for t := range doingTaskCh {
		t.wait()
	}
}
func (w *worker) run() {
	for {
		task , more := <- w.taskCh;
		result := doTask();
		task.result = result;
		task.Done()
		availableWorkerCh <- w
	}
}
```

### Check List 

Tests 

- No new test.

Code changes

Side effects

Related changes

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -966,112 +984,86 @@ func closeAddIndexWorkers(workers []*addIndexWorker) {
...
 // sendRangeTaskToWorkers sends tasks to workers, and returns remaining kvRanges that is not handled.
-func (w *worker) sendRangeTaskToWorkers(t table.Table, workers []*addIndexWorker, reorgInfo *reorgInfo, totalAddedCount *int64, kvRanges []kv.KeyRange) ([]kv.KeyRange, error) {
-	batchTasks := make([]*reorgIndexTask, 0, len(workers))
-	physicalTableID := reorgInfo.PhysicalTableID
-
+func sendRangeTaskToWorkers(ctx context.Context, t table.Table, reorgInfo *reorgInfo, kvRanges []kv.KeyRange, availableWorkerCh chan *addIndexWorker, doingTaskCh chan *reorgIndexTask) error {
 	// Build reorg indices tasks.
 	for _, keyRange := range kvRanges {
+		select {
+		case <-ctx.Done():
```
Comment (winkyao) [inline ddl/index.go]: Seems like this context will never be canceled.

File: ddl/index.go [hunk 2]
```diff
@@ -966,112 +984,86 @@ func closeAddIndexWorkers(workers []*addIndexWorker) {
...
-		batchTasks = append(batchTasks, task)
-
-		if len(batchTasks) >= len(workers) {
-			break
+		task := &reorgIndexTask{reorgIndexTaskData: reorgIndexTaskData{reorgInfo.PhysicalTableID, startHandle, endHandle, endIncluded}}
+		task.wg.Add(1)
+		var idxWorker *addIndexWorker
+		for {
+			idxWorker = <-availableWorkerCh
+			if !idxWorker.isClosed() {
```
Comment (winkyao) [inline ddl/index.go]:
```
if idxWorker.isClosed() {
	continue
}
```
is more readable

Comment (winkyao) [reply ddl/index.go]: Avoid double negative.

Comment (crazycs520) [reply ddl/index.go]: done.

File: ddl/index.go (hunk 1 above)
Comment (crazycs520) [reply ddl/index.go]: Maybe cancel in `waitTaskFinish`.

File: ddl/index.go [hunk 3]
```diff
@@ -1119,72 +1110,84 @@ func (w *worker) addPhysicalTableIndex(t table.PhysicalTable, indexInfo *model.I
...
 	idxWorkers := make([]*addIndexWorker, 0, workerCnt)
 	defer func() {
 		closeAddIndexWorkers(idxWorkers)
 	}()
+	availableWorkerCh := make(chan *addIndexWorker, variable.MaxDDLReorgWorkerCount) // max worker count
+	doingTaskCh := make(chan *reorgIndexTask, variable.MaxDDLReorgWorkerCount)       // max worker count
+	errCh := make(chan error, 1)
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		defer close(doingTaskCh)
```
Comment (winkyao) [inline ddl/index.go]: Wrap this go routine function to a function. make code clearer

Comment (winkyao) [reply ddl/index.go]: Maybe `dispatchTasks` e.g.

File: ddl/index.go [hunk 4]
```diff
@@ -1119,72 +1111,84 @@ func (w *worker) addPhysicalTableIndex(t table.PhysicalTable, indexInfo *model.I
...
+	workerCnt := int(variable.GetDDLReorgWorkerCounter())
 	idxWorkers := make([]*addIndexWorker, 0, workerCnt)
 	defer func() {
 		closeAddIndexWorkers(idxWorkers)
 	}()
+	availableWorkerCh := make(chan *addIndexWorker, variable.MaxDDLReorgWorkerCount) // max worker count
+	doingTaskCh := make(chan *reorgIndexTask, variable.MaxDDLReorgWorkerCount)       // max worker count
+	errCh := make(chan error, 1)
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
```
Comment (winkyao) [inline ddl/index.go]: assign this goroutine function to a local variable, and start it, like:
```
dispatchTaskFunc := {
	....
}

go dispatchTaskFunc()
```

Make the code be clearer.

File: ddl/index.go [hunk 5]
```diff
@@ -966,112 +984,87 @@ func closeAddIndexWorkers(workers []*addIndexWorker) {
...
+		default:
 		}
-
-		if result.err != nil {
-			log.Warnf("[ddl-reorg] worker[%v] return err:%v", i, result.err)
+		task.wg.Wait()
+		// task.result should never be nil here.
+		if task.result == nil {
+			cancel()
+			return errors.Errorf("[ddl-reorg] backfill index task result is nil")
```
Comment (winkyao) [inline ddl/index.go]: Error info should not contain `[ddl-reorg]`, the logger will choose to print it.

File: ddl/index.go [hunk 6]
```diff
@@ -966,112 +984,87 @@ func closeAddIndexWorkers(workers []*addIndexWorker) {
...
 		}
-
-		if firstErr == nil {
-			*totalAddedCount += int64(result.addedCount)
-			addedCount += int64(result.addedCount)
-			nextHandle = result.nextHandle
+		// check result != nil
```
Comment (winkyao) [inline ddl/index.go]: What is `check result != nil` means?

Comment (crazycs520) [reply ddl/index.go]: check task.result is not nil.

File: ddl/index.go [hunk 7]
```diff
@@ -1119,72 +1111,84 @@ func (w *worker) addPhysicalTableIndex(t table.PhysicalTable, indexInfo *model.I
...
 
 	// variable.ddlReorgWorkerCounter can be modified by system variable "tidb_ddl_reorg_worker_cnt".
-	workerCnt := variable.GetDDLReorgWorkerCounter()
+	workerCnt := int(variable.GetDDLReorgWorkerCounter())
 	idxWorkers := make([]*addIndexWorker, 0, workerCnt)
 	defer func() {
 		closeAddIndexWorkers(idxWorkers)
 	}()
+	availableWorkerCh := make(chan *addIndexWorker, variable.MaxDDLReorgWorkerCount) // max worker count
+	doingTaskCh := make(chan *reorgIndexTask, variable.MaxDDLReorgWorkerCount)       // max worker count
```
Comment (winkyao) [inline ddl/index.go]: doingTaskCh --> handlingTaskQueue

File: ddl/index.go [hunk 8]
```diff
@@ -966,112 +984,87 @@ func closeAddIndexWorkers(workers []*addIndexWorker) {
...
-		firstErr   error
-	)
-	for i := 0; i < taskCnt; i++ {
-		worker := workers[i]
-		result := <-worker.resultCh
-		if firstErr == nil && result.err != nil {
-			firstErr = result.err
-			// We should wait all working workers exits, any way.
-			continue
+func (w *worker) waitTaskFinish(cancel context.CancelFunc, reorgInfo *reorgInfo, availableWorkerCh chan *addIndexWorker, doingTaskCh chan *reorgIndexTask, errCh chan error) error {
```
Comment (winkyao) [inline ddl/index.go]: Add comments here, explain how why we need to wait the handling task finished sequentially

**Comment by shenli** [comment]
Will this PR improve performance?

**Comment by crazycs520** [comment]
@shenli , Theoretically, This PR will reduce the add-index-worker waiting time, so it should be faster than before.

**Comment by tiancaiamao** [comment]
IMHO, current code is much too complex to maintain... @crazycs520

**Pull request closed without merging**